File: hrms/hr/report/request_access_health/request_access_health.js
// Copyright (c) 2026, Nastyworldwide-Dev and contributors
// License: GNU General Public License v3. See license.txt

// Read-only. Every linked user is probed the way Nadi's create form would file
// for them; only the refusals are listed. Nothing on this page writes.
frappe.query_reports["Request Access Health"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
		},
		{
			fieldname: "doctype",
			label: __("Request Type"),
			fieldtype: "Select",
			options: [
				"",
				"Leave Application",
				"Attendance Request",
				"Shift Request",
				"Compensatory Leave Request",
				"Expense Claim",
				"OT Request",
				"Replacement Leave Claim",
				"Shift Swap Request",
				"Remote Checkin Request",
				"Employee Issue",
			],
		},
	],

	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.fieldname === "refused_by" && data && data.refused_by) {
			const colour = { role: "red", user_permission: "orange", identity: "red" }[data.refused_by] || "blue";
			value = `<span class="indicator-pill ${colour}">${frappe.utils.escape_html(data.refused_by)}</span>`;
		}
		return value;
	},
};

File: hrms/hr/report/request_access_health/request_access_health.py
"""Request Access Health — script report (15 Sep 2026).

Owner's rule: we never ask an employee to run a diagnostic; the system finds
it itself. One row per (user, request doctype) that user may NOT create for
their own employee, probed the way Nadi's form files it — as that user, own
employee set, no company, no posting date — with the gate that refused
(`hrms.api.diagnose`) and a plain sentence. An empty report is the healthy
state. Nothing here writes.

HR Manager and System Manager only; fenced to the caller's companies the way
the other HR reports are (`hrms.utils.report_scope.fenced_companies`), and a
fenced caller asking for another company is refused, not shown an empty page.
"""

from __future__ import annotations

import logging

import frappe
from frappe import _

from hrms.utils.report_scope import fenced_companies
from hrms.utils.request_access import PWA_REQUEST_DOCTYPES, scan

logger = logging.getLogger(__name__)


def execute(filters=None):
	filters = frappe._dict(filters or {})
	companies = fenced_companies(filters.get("company"))
	doctypes = (filters.doctype,) if filters.get("doctype") else PWA_REQUEST_DOCTYPES
	logger.info("[request_access_health] execute fence=%s doctypes=%d", companies or "none", len(doctypes))
	rows = scan(companies or None, doctypes)
	return _columns(), rows


def _columns() -> list[dict]:
	return [
		{"fieldname": "user", "label": _("User"), "fieldtype": "Link", "options": "User", "width": 220},
		{
			"fieldname": "employee",
			"label": _("Employee"),
			"fieldtype": "Link",
			"options": "Employee",
			"width": 130,
		},
		{"fieldname": "employee_name", "label": _("Employee Name"), "fieldtype": "Data", "width": 180},
		{
			"fieldname": "company",
			"label": _("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"width": 150,
		},
		{"fieldname": "doctype", "label": _("Request Type"), "fieldtype": "Data", "width": 190},
		{"fieldname": "refused_by", "label": _("Refused By"), "fieldtype": "Data", "width": 200},
		{"fieldname": "why", "label": _("Why"), "fieldtype": "Small Text", "width": 500},
	]
